src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs

```diff
@@ -105,40 +109,40 @@ public TextAndVersion GetValue(LoadTextOptions options, CancellationToken cancel
             var recoverableText = (RecoverableText)_initialSourceOrRecoverableText;
             if (recoverableText.LoadTextOptions != options && recoverableText.InitialSource != null)
             {
+                var textAndVersion = useAsync
+                    ? await recoverableText.InitialSource.GetValueAsync(options, cancellationToken).ConfigureAwait(false)
+                    : recoverableText.InitialSource.GetValue(options, cancellationToken);
                 Interlocked.Exchange(
                     ref _initialSourceOrRecoverableText,
-                    new RecoverableText(recoverableText.InitialSource, recoverableText.InitialSource.GetValue(options, cancellationToken), options, _services));
+                    new RecoverableText(recoverableText.InitialSource, textAndVersion, options, _services));
             }
 
-            recoverableText = (RecoverableText)_initialSourceOrRecoverableText;
-            return recoverableText.ToTextAndVersion(recoverableText.GetValue(cancellationToken));
+            return (RecoverableText)_initialSourceOrRecoverableText;
         }
 
-        public async Task<TextAndVersion> GetValueAsync(LoadTextOptions options, CancellationToken cancellationToken)
+        public TextAndVersion GetValue(LoadTextOptions options, CancellationToken cancellationToken)
         {
-            if (_initialSourceOrRecoverableText is ITextAndVersionSource source)
-            {
-                // replace initial source with recoverable text if it hasn't been replaced already:
-                Interlocked.CompareExchange(
-                    ref _initialSourceOrRecoverableText,
-                    value: new RecoverableText(source, await source.GetValueAsync(options, cancellationToken).ConfigureAwait(false), options, _services),
-                    comparand: source);
-            }
+#pragma warning disable CA2012 // Use ValueTasks correctly
+            var valueTask = GetRecoverableTextAsync(useAsync: false, options, cancellationToken);
+#pragma warning restore CA2012 // Use ValueTasks correctly
+            Contract.ThrowIfFalse(valueTask.IsCompleted, "GetRecoverableTextAsync should have completed synchronously since we passed 'useAsync: false'");
+            var recoverableText = valueTask.GetAwaiter().GetResult();
```

Comment:
sync path calls the helper with `useAsync: false` and knows then that it is completely safe to block on the result which must be completed.